src/metabase/driver/generic_sql/native.clj

```diff
@@ -16,37 +17,37 @@
-  (try (let [database (sel :one [Database :engine :details] :id database-id)
-             db (-> database
-                    db->korma-db
-                    korma.db/get-connection)
-             [columns & [first-row :as rows]] (jdbc/with-db-transaction [conn db :read-only? true]
-                                                ;; If timezone is specified in the Query and the driver supports setting the timezone
-                                                ;; then execute SQL to set it
-                                                (when-let [timezone (or (-> query :native :timezone)
-                                                                        (driver/report-timezone))]
-                                                  (when (seq timezone)
-                                                    (let [{:keys [features timezone->set-timezone-sql]} (driver/engine->driver (:engine database))]
-                                                      (when (contains? features :set-timezone)
-                                                        (log/debug "Setting timezone to:" timezone)
-                                                        (jdbc/db-do-prepared conn (timezone->set-timezone-sql timezone))))))
-                                                (jdbc/query conn sql :as-arrays? true))]
-         ;; TODO - Why don't we just use annotate?
-         {:rows    rows
-          :columns columns
-          :cols    (map (fn [column first-value]
-                          {:name      column
-                           :base_type (value->base-type first-value)})
-                        columns first-row)})
+  (try (let [database                            (sel :one :fields [Database :engine :details] :id database-id)
+             db-conn                             (-> database
+                                                     db->korma-db
+                                                     korma.db/get-connection)
+             {:keys [features set-timezone-sql]} (driver/engine->driver (:engine database))]
+
+         (jdbc/with-db-transaction [t-conn db-conn]
+
+           ;; Set the timezone if applicable. We do this *before* making the transaction read-only because some DBs
+           ;; won't let you set the timezone on a read-only connection
+           (when-let [timezone (driver/report-timezone)]
+             (when (and (seq timezone)
+                        (contains? features :set-timezone))
+               (log/debug (u/format-color 'green "%s" set-timezone-sql))
+               (try (jdbc/db-do-prepared t-conn set-timezone-sql [timezone])
+                    (catch Throwable e
+                      (log/error (u/format-color 'red "Failed to set timezone: %s" (.getMessage e)))))))
+
+           ;; Now make the transaction read-only and run the query itself
+           (.setReadOnly ^com.mchange.v2.c3p0.impl.NewProxyConnection (:connection t-conn) true)
```

Comment:
We want to use the connection pool the korma sets up for us, otherwise the connection will close as soon as the query is done executing and we won't be able to convert things like `PGobjects` to strings in JSON serialization.